src/Compilers/CSharp/Portable/Symbols/BaseTypeAnalysis.cs

```diff
@@ -105,30 +105,38 @@ private static void StructDependsClosure(NamedTypeSymbol type, HashSet<Symbol> p
         /// be managed even if it had no fields.  e.g. struct S { S s; } is not managed, but struct S { S s; object o; }
         /// is because we can point to object.
         /// </summary>
-        internal static bool IsManagedType(NamedTypeSymbol type)
+        internal static ManagedKind GetManagedKind(NamedTypeSymbol type)
         {
-            // If this is a type with an obvious answer, return quickly.
-            switch (IsManagedTypeHelper(type))
+            var hasGenerics = false;
+            var isManaged = IsManagedTypeHelper(type, ref hasGenerics);
```

Comment:
I didn't really want to solve this with a ref parameter, but propagating whether we saw a generic field during traversal without complicating the logic of deciding whether to continue searching was tricky.

Reply:
I think you can remove the parameter from IsManagedTypeHelper and just move the check into DependsOnDefinitelyManagedType. That'll at least bring it to one ref parameter.

Reply:
Also consider using a tuple return instead of ref return if the ref bothers you. i'm actually fine with the ref here but tuple would also be appropriate. Besides there is some fun in solving a tuple bug with a tuple :smile: 

---
In reply to: [245846507](https://github.com/dotnet/roslyn/pull/31602#discussion_r245846507) [](ancestors = 245846507)

Reply:
I may take another shot at converting it to return `ManagedKind` instead of `bool`. `ManagedKind.Managed` should stop the traversal while `ManagedKind.UnmanagedWithGenerics` should continue searching for a managed field while propagating up the stack that we saw a generic field.